### ISSUE TITLE
World caching problems

### ISSUE THREAD

**Issue by rob9315** [rob9315/mcproxy#2]
(originally by @etiaro)
> Hi,
I'm experiencing bugs with chunks caching:
blank textures: I found this one pretty interesting, as client knows where are blocks(and their types), just refuses to show textures for some reason. Sometimes enabling and disabling x-ray shows the textures, but when go to neighbour chunk and look away, the corrupted chunk becomes blank again. Also, mods like chest-ESP are showing positions correctly
> 
> Shifting chunks: on 2b2t's spawn, or the nether(so probably on chunks filled with much blocks), some of the chunks are getting shifted down. Seems like always y is subtracted by multiply of some number(I didn't count tbh). It's best visible in the nether, as the roof is flat bedrock. Almost all of broken chunk's roofs are on the same level, and a few looks like they've got shifted twice. Sometimes there's just one corrupted chunk, and sometimes there is whole line(2x8 was largest i saw), in some cases they're completely random, I can't see any pattern there.
>
> I am running proxy on really low-cpu and low-RAM(1gb ram raspberry pi, arch linux) so maybe it somehow causes the problem.

[answer by @rob9315]
> > blank textures: I found this one pretty interesting, as client knows where are blocks(and their types), just refuses to show textures for some reason. Sometimes enabling and disabling x-ray shows the textures, but when go to neighbour chunk and look away, the corrupted chunk becomes blank again. Also, mods like chest-ESP are showing positions correctly
> 
> Ah yes, I know this one. This happens because of the way [mineflayer](https://github.com/PrismarineJS/mineflayer) calculates/caches light information, which gets out of sync whenever you change something and relog.
> > Shifting chunks: on 2b2t's spawn, or the nether(so probably on chunks filled with much blocks), some of the chunks are getting shifted down. Seems like always y is subtracted by multiply of some number(I didn't count tbh). It's best visible in the nether, as the roof is flat bedrock. Almost all of broken chunk's roofs are on the same level, and a few looks like they've got shifted twice. Sometimes there's just one corrupted chunk, and sometimes there is whole line(2x8 was largest i saw), in some cases they're completely random, I can't see any pattern there.
> 
> I have no clue though, what causes this.
> > I am running proxy on really low-cpu and low-RAM(1gb ram raspberry pi, arch linux) so maybe it somehow causes the problem.
> 
> I don't think that has anything to do with it, but good that you use arch BTW 1st_place_medal
> 
> TLDR: The lighting issue can be fixed, the PrismarineJS team was working on that already at some point, breaking a block should also fix it. The chunk shifting problem i have no clue why it happens or how to fix it, sorry.

[answer by @MrGeorgen]
> related issues to the lighting problem:
PrismarineJS/mineflayer#591
PrismarineJS/prismarine-world#46

**Comment by czho** [comment]
is this lighting only cause theres a weird issue where it swaps part of the world sometimes

**Comment by etiaro** [comment]
Actually, it doesn't happen in regular overworld, when i got out of spawn, the only place where chunks 'shift' is the nether

**Comment by rob9315** [comment]
That is still extremely weird, maybe an inaccuracy with high coordinate values? How far out did this happen approximately?

**Comment by czho** [comment]
> Actually, it doesn't happen in regular overworld, when i got out of spawn, the only place where chunks 'shift' is the nether

it does do this some times at spawn, the issue might be with either too much chunk data or being too high in the air

**Comment by rob9315** [comment]
Does this happen while being connected to the proxy or only when reconnecting?

**Comment by etiaro** [comment]
For me it's just while connecting. After changing world it's fixed

**Comment by czho** [comment]
yea, its an issue with chunks that are cached

**Comment by czho** [comment]
alright I looked into it a bit and basically the y level gets messed u sometimes and thats alot of the issues

**Comment by rob9315** [comment]
Please try again with the newest beta, with the help of Kami I think I managed to fix the issue which was there because the `map_chunk` packets could be too large with too much data.

**Comment by rob9315** [comment]
Has been fixed with the latest commit
